### PR TITLE
persist: relax the requirements on allow_compaction and seal

### PR DESCRIPTION
**Pull request by ruchirK**
Previously, we required that all allow compaction calls be greater than the current
since frontier. However, because we instantiated each trace with a since frontier
of 0, that meant every user has to know to not try to "advance" the since frontier
to zero.

Unfortunately, this is different from the contract differential arrangments expose,
where it is safe to allow compactions to 0 (or more generally T::minimum) so for now
we will relax this invariant as it was never really clear why had it in the first
place.

### CODE REVIEW

**Comment by ruchirK** [comment]
@danhhz I don't have a strong thought about the invariant but I'm more concerned that now when we boot up the first thing you see in the log is "compaction failed 0 less than or equal to trace since 0" 

I felt like we could
- change coord to not send 0 frontiers or change persist to ignore 0 frontiers
- relax the invariant
- change the log severity

and of those option #2 seemed the best. I double checked differential's take here

`set_logical_compaction` on `TraceReader` says:

```
It is an error to call this method with a frontier not beyond the most recent arguments to this method, or the initial value of get_logical_compaction() if this method has not yet been called.
```
which sounds most like option 1 (the initial value is always T::minimum)

from: https://docs.rs/differential-dataflow/0.12.0/differential_dataflow/trace/trait.TraceReader.html#tymethod.set_logical_compaction

But if you look at an actual impl it looks like:

```
    fn set_logical_compaction(&mut self, frontier: AntichainRef<T>) {
        // TODO: Re-use allocation
        self.logical_frontier = frontier.to_owned();
    }
```

so it doesn't seem like its the highest priority :)

**Review by danhhz — APPROVED** [review]
What concretely prompted this? Just maintaining behavioral parity with dd or did a test or something fail?

Also, is there any good reason seal has to error on ==? I've always found that awkward and don't remember why we disallow it.

**Comment by ruchirK** [comment]
> What concretely prompted this?

i think i answered this as you were reviewing!

**Comment by danhhz** [comment]
You did! Let's put that context in the commit message/PR description. While looking at code, it's much nicer to have that sort of thing at hand than having to follow down to the github PR comments :)

I imagine seal was the same reason (matching dd). I'm inclined to have them match each other, over having one match dd and the other not

**Comment by ruchirK** [comment]
sure! so thats "teach persist to ignore allow_compactions where the since frontier equals zero" right?

**Comment by ruchirK** [comment]
and yeah seal was to match the invariants around capability downgrades (which are more strongly enforced iirc)

**Comment by danhhz** [comment]
> sure! so thats "teach persist to ignore allow_compactions where the since frontier equals zero" right?

not quite. that's still talking about what the change is, but i'm trying to get at why make the change at all. something like "fix unnecessary logged error being printed on every startup after 7966 merged: compaction failed 0 less than or equal to trace since 0"

**Comment by ruchirK** [comment]
sorry sorry I should have been more clear.

I understood the point about being more forthcoming about why I'm making the change in the commit message. I was specifically asking about what change we ought to make.


BUT capability downgrades actually allow >= https://docs.rs/timely/0.12.0/timely/dataflow/operators/struct.Capability.html#method.downgrade

chatted with frank and the difference between the two is for good reason. We're going to talk about this on Monday to make sure we understand why. Ill mark this as a draft till then

**Comment by ruchirK** [comment]
Spoke with @frankmcsherry about this and the takeaways were:

- For both capability downgrades and set_{logical,physical}_compaction in timely / differential respective, the appropriate restriction is "arguments must be >= the previously sent argument, or equal to the initial value"
- the persistence crate's seal and allow_compaction commands map directly to those concepts and should have that restriction
- someone (probably me) needs to change the text in the differential repo and sanity check that the surrounding code doesn't break through e.g. an over-aggressive assert.

The second and third bullets can be done in parallel. I don't think we need to wait for differential's update to get merged before updating the semantics in the persistence crate.

**Comment by ruchirK** [comment]
https://github.com/TimelyDataflow/differential-dataflow/pull/339 is the corresponding change to differential

**Review by danhhz — APPROVED** [review]

**Pull request merged**
